src/validation.cpp

```diff
@@ -2630,6 +2627,9 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
         }
 
         if (nStopAtHeight && pindexNewTip && pindexNewTip->nHeight >= nStopAtHeight) StartShutdown();
+
+        if (ShutdownRequested())
```

Comment:
In commit "Fix fast-shutdown crash if genesis block was not loaded"

Would add short comment here like "Only allow shutting down *after* calling ActivateBestChainStep to avoid problems with shutdown code assuming there is a known best block."

It'd seem cleaner to just fix the broken flush code (and also probably easier to write a unit test for), but I realise you are trying to implement a quick fix.